Kudu.Core/SourceControl/Git/LibGit2SharpRepository.cs

```diff
@@ -0,0 +1,396 @@
+                        b => b.TrackedBranch = trackedBranch.CanonicalName);
+
+                    //Update the raw ref to point the head of branchName to the latest fetched branch
+                    UpdateRawRef(string.Format("refs/heads/{0}", branchName), trackedBranchName);
+
+                    // Now checkout out our branch, which points to the right place
+                    Update(branchName);
+                }
+            }
+            finally
+            {
+                using (var repo = new LibGit2Sharp.Repository(RepositoryPath))
+                {
+                    repo.Network.Remotes.Remove(_remoteAlias);
+                }
+            }
+        }
+
+        public void Clean()
+        {
+            using (var repo = new LibGit2Sharp.Repository(RepositoryPath))
+            {
+                repo.RemoveUntrackedFiles();
+            }
+        }
+
+        public void CreateOrResetBranch(string branchName, string startPoint)
+        {
+            using (var repo = new LibGit2Sharp.Repository(RepositoryPath))
+            {
+                var branch = repo.Branches[branchName];
+                if (branch == null)
+                {
+                    branch = repo.CreateBranch(branchName, startPoint);
+                    repo.Checkout(branch);
+                }
+                else if (string.IsNullOrEmpty(startPoint))
+                {
+                    repo.Checkout(branch);
+                }
+                else
+                {
+                    var commit = repo.Lookup<Commit>(startPoint);
+                    if (commit != null)
+                    {
+                        repo.Checkout(branch);
+                        repo.Reset(ResetMode.Hard, commit);
+                    }
+                }
+            }
+        }
+
+        public void UpdateRef(string toBeUpdatedToAlias)
+        {
+            UpdateRawRef(string.Format("refs/heads/{0}", "master"), string.Format("refs/heads/{0}", toBeUpdatedToAlias));
+        }
+
+        public void UpdateRawRef(string toBeUpdatedRef, string toBeUpdatedToRef)
+        {
+            using (var repo = new LibGit2Sharp.Repository(RepositoryPath))
+            {
+                repo.Refs.UpdateTarget(toBeUpdatedRef, toBeUpdatedToRef);
+            }
+        }
+
+        public bool DoesBranchContainCommit(string branchName, string commitOrBranchName)
+        {
+            using (var repo = new LibGit2Sharp.Repository(RepositoryPath))
+            {
+                var branch = repo.Branches[branchName];
+                if (branch == null) return false;
+                return repo.Refs.ReachableFrom(
+                              new [] {repo.Refs[branch.CanonicalName]},
+                              new [] {repo.Lookup<Commit>(commitOrBranchName)}).Any();
+            }
+        }
+
+        public IEnumerable<string> ListFiles(string path, SearchOption searchOption, params string[] lookupList)
+        {
```

Comment:
@suwatch Would you have some kind of weighted list of calls against the current api based on feedback from production?

That may make sense to reproduce this distribution, then (eventually) focus on trying to speed up the "hot paths"